test_runner.py

```diff
@@ -0,0 +1,160 @@
+            for s in socks:
+                # When binding a socket to port 0, the kernel will assign it a free port
+                s.bind(('', 0))
+            ports = [str(s.getsockname()[1]) for s in socks]
+            for s in socks:
+                # Queue up to 10 requests
+                s.listen(10)
+            return ports
+
+        rest_port, http_port = get_free_ports(2)
+        instance = 'http://rest-server:%s' % rest_port
+        print('Creating another CodaLab instance {} at {} for testing...'.format(name, instance))
+
+        try:
+            start_time = time.time()
+            subprocess.check_call(
+                ' '.join(
+                    [
+                        'python3',
+                        TestRunner._CODALAB_SERVICE_SCRIPT,
+                        'start',
+                        '--instance-name %s' % name,
+                        '--rest-port %s' % rest_port,
+                        '--version %s' % version,
+                        '--services init rest-server',
+                    ]
+                ),
+                shell=True,
+            )
+            print(
+                'It took {} seconds to create the temp instance.'.format(time.time() - start_time)
+            )
+        except subprocess.CalledProcessError as ex:
+            print('There was an error while creating the temp instance: %s' % ex.output)
+            raise
+
+        return instance
+
+    def __init__(self, args):
+        self.instance = args.instance
+        self.tests = args.tests
+
+        # Check if a second, temporary instance of CodaLab is needed for testing
+        self.temp_instance_required = any(
+            test in self.tests for test in TestRunner._TEMP_INSTANCE_NEEDED_TESTS
+        )
+        if self.temp_instance_required:
+            self.temp_instance_name = 'temp-instance%s' % ''.join(
+                random.choice(string.digits) for _ in range(8)
+            )
+            self.temp_instance = TestRunner._create_temp_instance(
+                self.temp_instance_name, args.version
+            )
+
+    def run(self):
+        success = True
+        try:
+            # Run backend tests using test_cli
+            test_command = 'python3 test_cli.py --instance %s ' % self.instance
+            if self.temp_instance_required:
+                test_command += '--second-instance %s ' % self.temp_instance
+            test_command += ' '.join(list(filter(lambda test: test != 'frontend', self.tests)))
+
+            print('Running backend tests with command: %s' % test_command)
+            subprocess.check_call(TestRunner._docker_exec(test_command), shell=True)
+
+            # Run frontend tests
+            self._run_frontend_tests()
+
+        except subprocess.CalledProcessError as ex:
+            print('Exception while executing tests: %s' % ex.output)
+            success = False
+
+        self._cleanup()
+        return success
+
+    def _run_frontend_tests(self):
```

Comment:
I run the Selenium tests here now too as long as the `frontend` module key is passed in.